tools/nacl_helper_bootstrap-armhf/nacl_helper_bootstrap-armhf.c

```diff
@@ -0,0 +1,67 @@
+/*
+===========================================================================
+Daemon BSD Source Code
+Copyright (c) 2022, Daemon Developers
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the Daemon developers nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h> 
+
+extern char **environ;
+
+int main(int argc, char *argv[]) {
+	char *wrapper = "nacl_helper_bootstrap-armhf";
+	char *directory = argv[0];
+	directory[strlen(directory) - strlen(wrapper)] = '\0';
```

Comment:
This has no consequence AFAIK, but you are modifying argv here :eyes:

Reply:
Yes… it's a bit tricky, I extract the dirpath from the fullpath by writing a `\0` in place of the file name… 👀️

Also the code is not C++ compliant (that's why I made it a C code 😜️): I affect `const char*` to `char*`. when setting `wrapper`, `helper` and doing `putenv`. 🤷‍♀️️ I'm not sure we have to over-engineer this tiny code. 😏️

Reply:
huh oh. Yeah, so how about doing `char *directory = strdup(argv[0]);` and make argv a `const char*` instead? you don't even need to care about freeing memory because the process exits afterwards.

Reply:
Oh by the way, if we are going to use the libc, you can also do that:
```C
#include <libgen.h>
[…]
char *directory = dirname(strdup(argv[0]);
```

Reply:
How about like this to avoid the hackiest assumptions about argv[0]:

```
char* slash = strrchr(argv[0], '/');
if (slash) {
    *slash = '/0';
    int err = chdir(argv[0]);
    if (err != 0) { return err; }
}
```

Reply:
The reason why I made this as simple is that it really has only one usage in the world and the only program expected to run it (the engine) has its name hardcoded anyway. The bash equivalent script was just 2 lines…

If we're going to implement a real `dirname` maybe we should just use `dirname` like @necessarily-equal suggested…

Reply:
I applied @necessarily-equal's suggestion and turned the whole program into a c++ one, using `strdup` when setting `char*` from `const char*` constants.